File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let len = height.length;
    prefixMax = new Array(len).fill(0);
    prefixMax[0] = height[0];
    suffixMax = new Array(len).fill(0);
    suffixMax[len - 1] = height[len - 1];

    for(let i = 1; i < len; i++) {
        prefixMax[i] = Math.max(prefixMax[i-1], height[i]);
    }
    for(let i = len-2; i >= 0; i--) {
        suffixMax[i] = Math.max(suffixMax[i+1], height[i]);
    }
    
    let totalWater = 0;
    for(let i = 0; i < len; i++) {
        totalWater += Math.min(prefixMax[i], suffixMax[i]) - height[i];
    }
    return totalWater;
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        total = 0
        prefix_max = [0 for _ in range(n)] 
        prefix_max[0] = height[0]
        
        suffix_max = [0 for _ in range(n)] 
        suffix_max[n-1] = height[n-1]
        
        for i in range(1, n):
            prefix_max[i] = max(prefix_max[i-1], height[i])
        for i in range(n-2, -1, -1):
            suffix_max[i] = max(suffix_max[i+1], height[i])

        for i in range(0, n):
            total += min(prefix_max[i], suffix_max[i]) - height[i]
        return total 